### PR TITLE
enable 66839 in 4.15 upgrade

### PR DESCRIPTION
**Pull request by jiajliu**
Follow up pr of https://github.com/openshift/release/pull/43227 to handle the 4.14-4.15 upgrade
1. add CAPABILITIES_IN_CREDENTIALREQUEST to distinguish different caps with credential requests in different version
2. add machineapi additional cap for 4.14 aws installation in 4.14-4.15 upgrade

### CODE REVIEW

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-4.14-upgrade-from-stable-4.13-aws-ipi-disconnected-sts-basecap-none-f28

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disconnected-sts-basecap-none-f28

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disconnected-sts-basecap-none-f28

**Comment by jiajliu** [comment]
@jianlinliu @shellyyang1989 ptal, thanks.

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-ref.yaml [hunk 1]
```diff
@@ -15,6 +15,14 @@ ref:
     default: "false"
     documentation: |-
       Detemine if OTA test will be included
+  - name: CAPABILITIES_IN_CREDENTIALREQUEST
+    default: ""
+    documentation: |-
+      This is a variable uesd for OTA test. It has different values in different versions. It represents the capabilities with capability annotation in CredentialsRequest.
+  - name: ADDITIONAL_ENABLED_CAPABILITIES
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-ref.yaml:22]: In postupgrade step, seem like this var is not needed.

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly-4.14-upgrade-from-stable-4.13.yaml [hunk 2]
```diff
@@ -191,6 +191,7 @@ tests:
     env:
       BASE_DOMAIN: qe.devcluster.openshift.com
       BASELINE_CAPABILITY_SET: None
+      CAPABILITIES_IN_CREDENTIALREQUEST: ImageRegistry MachineAPI
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly-4.14-upgrade-from-stable-4.13.yaml:194]: How about rename the  var to `EXPECTED_CAPABILITIES_IN_CREDENTIALREQUEST` for easy reading.

File: ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh [hunk 3]
```diff
@@ -35,12 +35,14 @@ function pre-ocp-66839(){
...
         return 1
     fi
-    # There should not be any cap annotation in all extracted manifests 
-    curCap=$(grep -rh "capability.openshift.io/name:" "${manifestsDir}"|awk -F": " '{print $NF}'|sort -u)
-    if [[ "${curCap}" != "" ]]; then
-        echo "Caps in extracted manifests found: ${curCap}, but expected nothing"
+
+    # There should be only enabled cap annotaion in all extracted manifests
+    curCap=$(grep -rh "capability.openshift.io/name:" "${manifestsDir}"|awk -F": " '{print $NF}'|sort -u|xargs)
+    if [[ "${curCap}" != "${ADDITIONAL_ENABLED_CAPABILITIES}" ]]; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh:41]: Per the logic, `ADDITIONAL_ENABLED_CAPABILITIES` has to be in a sorted order for multiple caps, while it is also used in fresh install, while in fresh install, there is no a rule to limit its value has to be in a sorted order. 

To make the codes be robust, maybe we need to sort `ADDITIONAL_ENABLED_CAPABILITIES` before the compare? Similar comments on https://github.com/openshift/release/pull/44820/files#diff-bcec571809d92fb0e609baf8907886244e300b0f2aa08759cc9f6cac2154a225R39

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-ref.yaml (hunk 1 above)
Comment (jiajliu) [reply ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-ref.yaml]: removed

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly-4.14-upgrade-from-stable-4.13.yaml (hunk 2 above)
Comment (jiajliu) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly-4.14-upgrade-from-stable-4.13.yaml:194]: updated

File: ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh (hunk 3 above)
Comment (jiajliu) [reply ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh]: good point, updated

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disconnected-sts-basecap-none-f28

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disconnected-sts-basecap-none-f28

**Comment by jiajliu** [comment]
@jianlinliu PTAL agains, thanks!

File: ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh [hunk 4]
```diff
@@ -35,12 +35,15 @@ function pre-ocp-66839(){
...
         return 1
     fi
-    # There should not be any cap annotation in all extracted manifests 
-    curCap=$(grep -rh "capability.openshift.io/name:" "${manifestsDir}"|awk -F": " '{print $NF}'|sort -u)
-    if [[ "${curCap}" != "" ]]; then
-        echo "Caps in extracted manifests found: ${curCap}, but expected nothing"
+
+    # There should be only enabled cap annotaion in all extracted manifests
+    curCap=$(grep -rh "capability.openshift.io/name:" "${manifestsDir}"|awk -F": " '{print $NF}'|sort -u|xargs)
+    expectedCap=$(echo ${ADDITIONAL_ENABLED_CAPABILITIES} | tr ' ' '\n' | sort)
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh:41]: `curCap` is a one-line string, `expectedCap` is a multi-lines string, right?

Comment (jiajliu) [reply ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh:41]:
```
# cat try.sh 
ADDITIONAL_ENABLED_CAPABILITIES="MachineAPI ImageRegistry"
test=$(echo ${ADDITIONAL_ENABLED_CAPABILITIES} | tr ' ' '\n' | sort)
echo ${test}
# ./try.sh 
ImageRegistry MachineAPI
```
Also one-line.

Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh:41]:
```
$ cat test.sh 
ADDITIONAL_ENABLED_CAPABILITIES="MachineAPI ImageRegistry"
curCap=$(echo ${ADDITIONAL_ENABLED_CAPABILITIES} |awk -F": " '{print $NF}'|sort -u|xargs)
expectedCap=$(echo ${ADDITIONAL_ENABLED_CAPABILITIES} | tr ' ' '\n' | sort)

if [[ "${curCap}" != "${expectedCap}" ]]; then echo "failed"; fi

$ sh -x test.sh 
+ ADDITIONAL_ENABLED_CAPABILITIES='MachineAPI ImageRegistry'
++ echo MachineAPI ImageRegistry
++ awk '-F: ' '{print $NF}'
++ sort -u
++ xargs
+ curCap='MachineAPI ImageRegistry'
++ echo MachineAPI ImageRegistry
++ tr ' ' '\n'
++ sort
+ expectedCap='ImageRegistry
MachineAPI'
+ [[ MachineAPI ImageRegistry != \I\m\a\g\e\R\e\g\i\s\t\r\y\
\M\a\c\h\i\n\e\A\P\I ]]
+ echo failed
failed
```

If I changed `expectedCap=$(echo ${ADDITIONAL_ENABLED_CAPABILITIES} | tr ' ' '\n' | sort)` to `expectedCap=$(echo ${ADDITIONAL_ENABLED_CAPABILITIES} | sort -u | xargs)`, it works.

Comment (jiajliu) [reply ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh:41]: You're right. we need xargs to change it to one-line. Let me update.

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh [hunk 5]
```diff
@@ -36,8 +36,8 @@ function post-ocp-66839(){
     fi
     # New gained cap annotation should be in extracted creds 
     newCap=$(grep -rh "capability.openshift.io/name:" "${credsDir}"|awk -F": " '{print $NF}'|sort -u|xargs)
-    if [[ "${newCap}" != "MachineAPI ImageRegistry" ]] && [[ "${newCap}" != "ImageRegistry MachineAPI" ]]; then
-        echo "Tobe gained CRs with cap annotation: ${newCap}, but expected: MachineAPI and ImageRegistry"
+    if [[ "${newCap}" != "${EXPECTED_CAPABILITIES_IN_CREDENTIALREQUEST}" ]]; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:39]: can we do some similar sort for `EXPECTED_CAPABILITIES_IN_CREDENTIALREQUEST` before the comparision.

File: ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh [hunk 6]
```diff
@@ -64,18 +67,28 @@ function pre-ocp-66839(){
...
+
     # Extract all CRs from tobe upgrade release payload with --included
     if ! oc adm release extract --to "${tobeCredsDir}" --included --credentials-requests "${OPENSHIFT_UPGRADE_RELEASE_IMAGE_OVERRIDE}"; then
         echo "Failed to extract CRs from tobe upgrade release payload!"
         return 1
     fi
     tobecap=$(grep -rh "capability.openshift.io/name:" "${tobeCredsDir}"|awk -F": " '{print $NF}'|sort -u|xargs)
-    if [[ "${tobecap}" != "MachineAPI ImageRegistry" ]] && [[ "${tobecap}" != "ImageRegistry MachineAPI" ]]; then
-        echo "Tobe gained CRs with cap annotation: ${tobecap}, but expected: MachineAPI and ImageRegistry"
+    if [[ "${tobecap}" != "${EXPECTED_CAPABILITIES_IN_CREDENTIALREQUEST}" ]]; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh:90]: the same comments as https://github.com/openshift/release/pull/44820/files#r1377095365

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh (hunk 5 above)
Comment (jiajliu) [reply ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh]: updated

**Comment by jianlinliu** [comment]
cc @liangxia to review.

**Pull request merged**